File: src/test/java/de/lukaseichler/recurrentneuralnetwork/NodeTest.java
package de.lukaseichler.recurrentneuralnetwork;

import java.util.Arrays;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;
import com.google.common.collect.Lists;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * @author leichler
 */
public class NodeTest extends NeuralNetworkTest {

    private Node node;

    @BeforeMethod
    public void setUp() throws Exception {
        node = new Node();
    }

    @Test
    public void inputCalculated() throws Exception {
        assertThat(node.calculate(Lists.newArrayList(1.5, 2.0))).isEqualTo(applyDefaultActivationFunction(3.5));
    }

    @Test(expectedExceptions = NullPointerException.class)
    public void nullActivationFunction() throws Exception {
        new Node(null);
    }

    @Test
    public void nodeWithActivationFunction() throws Exception {
        node.setActivationFunction(new LogActivation());
        assertThat(node.calculate(Lists.newArrayList(1.0, 2.0))).isEqualTo(applyDefaultActivationFunction(3));
    }

    @Test
    public void logAsDefaultActivationFunction() throws Exception {
        assertThat(node.getActivationFunction()).isInstanceOf(LogActivation.class);
    }

    @Test
    public void weightedInput() throws Exception {
        node.setWeights(2, 3);
        assertThat(node.calculate(Lists.newArrayList(1.0, 1.0))).isEqualTo(applyDefaultActivationFunction(5));
    }


    @Test
    public void weightedInputWithActivationFunction() throws Exception {
        node.setWeights(2, 3);
        node.setActivationFunction(new LogActivation());
        assertThat(node.calculate(Lists.newArrayList(1.0, 1.0))).isEqualTo(applyDefaultActivationFunction(5));
    }

    @Test
    public void weightUpdate() throws Exception {
        node.updateWeights(Arrays.asList(0.3));
        assertThat(node.calculate(Arrays.asList(1.0))).isEqualTo(applyDefaultActivationFunction(1.3));
    }

    @Test
    public void weightsUpdateReturnsUpdatedWeights() throws Exception {
        assertThat(node.updateWeights(Arrays.asList(0.3, 0.5))).containsAll(Arrays.asList(1.3, 1.5));
    }

    @Test(expectedExceptions = NullPointerException.class)
    public void weightUpdateNullDeltas() throws Exception {
        node.updateWeights(null);
    }

    @Test(expectedExceptions = IllegalArgumentException.class)
    public void mismatchingWeightsAndDeltas() throws Exception {
        node.calculate(Arrays.asList(1.0, 1.0));
        node.updateWeights(Arrays.asList(1.0));
    }

}

File: src/test/scala/de/lukaseichler/recurrentneuralnetwork/structure/NodeTest.scala
package de.lukaseichler.recurrentneuralnetwork.structure

import de.lukaseichler.recurrentneuralnetwork.NeuralNetworkTest
import de.lukaseichler.recurrentneuralnetwork.activation.LogActivation
import org.assertj.core.api.Assertions._
import org.scalatest.Matchers
import org.scalatest.testng.TestNGSuite
import org.testng.annotations.{BeforeMethod, Test}

/**
 * @author Lukas Eichler
 */
class NodeTest extends TestNGSuite with Matchers with NeuralNetworkTest {
  private var node: Node = null

  @BeforeMethod
  @throws(classOf[Exception])
  def setUp {
    node = new Node
  }

  @Test
  @throws(classOf[Exception])
  def inputCalculated {
    assertThat(node.calculate(List(1.5, 2.0))).isEqualTo(applyDefaultActivationFunction(3.5))
  }

  @Test(expectedExceptions = Array(classOf[IllegalArgumentException]))
  @throws(classOf[Exception])
  def nullActivationFunction {
    new Node(null)
  }

  @Test
  @throws(classOf[Exception])
  def nodeWithActivationFunction {
    node.setActivationFunction(new LogActivation)
    assertThat(node.calculate(List(1.0, 2.0))).isEqualTo(applyDefaultActivationFunction(3))
  }

  @Test
  @throws(classOf[Exception])
  def logAsDefaultActivationFunction {
    node.getActivationFunction.isInstanceOf[LogActivation]
  }

  @Test
  @throws(classOf[Exception])
  def weightedInput {
    node.setWeights(2, 3)
    assertThat(node.calculate(List(1.0, 1.0))).isEqualTo(applyDefaultActivationFunction(5))
  }

  @Test
  @throws(classOf[Exception])
  def weightedInputWithActivationFunction {
    node.setWeights(2, 3)
    node.setActivationFunction(new LogActivation)
    assertThat(node.calculate(List(1.0, 1.0))).isEqualTo(applyDefaultActivationFunction(5))
  }

  @Test
  @throws(classOf[Exception])
  def weightUpdate {
    node.updateWeights(List(0.3))
    assertThat(node.calculate(List(1.0))).isEqualTo(applyDefaultActivationFunction(1.3))
  }

  @Test
  @throws(classOf[Exception])
  def weightsUpdateReturnsUpdatedWeights {
    node.updateWeights(List(0.3, 0.5)) should (contain(1.3) and contain(1.5))
  }

  @Test(expectedExceptions = Array(classOf[IllegalArgumentException]))
  @throws(classOf[Exception])
  def weightUpdateNullDeltas {
    node.updateWeights(null)
  }

  @Test(expectedExceptions = Array(classOf[IllegalArgumentException]))
  @throws(classOf[Exception])
  def mismatchingWeightsAndDeltas {
    node.calculate(List(1.0, 1.0))
    node.updateWeights(List(1.0))
  }
}
